torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -1255,23 +1257,32 @@ void ProcessGroupNCCL::workCleanupLoop() {
         }
 
         // Report desync state in case of timeout
-        if (desyncDebug_ && timedOut) {
+        if (timedOut) {
           try {
-            // Set shutdown mode, so the heartbeat monitor thread will not
-            // abort process immediately.
-            collectiveDebugInfoMode_.store(true);
-            // Store debug info to storage. (By default to local disk)
-            auto dumpingDebugInfo = tryWriteDebugInfo();
-            auto desyncMsg = getNCCLWatchdogDebugInfo();
-            LOG(ERROR) << desyncMsg;
-            if (dumpingDebugInfo && dumpingDebugInfo->joinable()) {
-              std::this_thread::sleep_for(
-                  std::chrono::milliseconds(kWatchdogThreadSleepMillis * 30));
-              // At this point, we either have already waited for
-              // `kWatchdogThreadSleepMillis * 30` or the thread has finished so
-              // that we mark the thread detach and the dump of debug info
-              // becomes "best effort".
-              dumpingDebugInfo->detach();
+            if (desyncDebug_ || dumpOnTimeout_) {
+              // Set shutdown mode, so the heartbeat monitor thread will not
+              // abort process immediately.
+              collectiveDebugInfoMode_.store(true);
+            }
+
+            c10::optional<std::thread> dumpingDebugInfo;
```

Comment:
Is this used outside the if below?

Reply:
oh, i did i this way at first so i could preserve having the sleep below the desync clause.

i can put it back that way if it is better. let me see if i can understand your comment above.

Reply:
Well, I don't know if we sleep before getting the desync report will lead to heartbeat thread kick in early to kill the watchdog so that desync report is not complete.